linode/manager
# fix: [M3-5783] Firewall Rules - Improve form requirements and feedback

## Description 📝
Cloud Firewall rules drawer allowed custom rules to be saved without defining a "source" which causes the rule to fail.
It also allowed to save without a label or ports, which isn't correct.

this PR:
- Label the required fields as "Required" (Label, Protocol, Ports, and Sources)
- Handles errors for new required fields
- Validate that a selection has been made on the"Sources" input when the "Add Rule" button is clicked
- Remove the logic that disables the button

**NOTE**: Apologies for the diff! it's harder to read because I split types, utilities and the form component into their own files for more clarity and readability. That being said this PR should be tested manually for the form validation patterns

## Preview 📷

| Before  | After   |
| ------- | ------- |
| ![Screenshot 2023-05-16 at 2 44 13 PM](https://github.com/linode/manager/assets/130582365/20ce4d61-7aa6-405e-9528-666a05248602) | ![Screenshot 2023-05-16 at 2 44 26 PM](https://github.com/linode/manager/assets/130582365/f300c218-6a3b-4d23-9ef7-9fb95be2b8ba) |

## How to test 🧪
1. Pull code locally, and make sure you have a Linode with an associated firewall
2. Go to /firewalls and click on one of your Firewall to configure it
3. Click on "Add an Inbound Rule"
- - Click on the "Add rule" button and make sure the required fields all have an error feedback
- - Toggle/Fill some of the fields and make sure the validation get updated when clicking on the  "Add rule" button again
- - Try various combinations to toggle the form state and make sure you are able to create a rule with the new constraints (ex: an `ICMP` or `IPENCAP` protocol should disable the requirement on the `Ports` field)
